src/desktop/dialogs/curvepresetdialog.cpp

```diff
@@ -0,0 +1,304 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "desktop/dialogs/curvepresetdialog.h"
+#include "desktop/widgets/kis_curve_widget.h"
+#include "libclient/utils/icon.h"
+#include <QGridLayout>
+#include <QHBoxLayout>
+#include <QInputDialog>
+#include <QListWidget>
+#include <QMessageBox>
+#include <QPushButton>
+#include <QSettings>
+#include <QtMath>
+
+namespace dialogs {
+
+CurvePresetDialog::CurvePresetDialog(
+	const KisCubicCurve &current, bool linear, QWidget *parent)
+	: QDialog(parent)
+{
+	setModal(true);
+
+	m_curveWidget = new KisCurveWidget{this};
+	m_curveWidget->setReadOnly(true);
+	m_curveWidget->setLinear(linear);
+	m_curveWidget->setFixedSize(300, 300);
+
+	m_presetList = new QListWidget{this};
+
+	QHBoxLayout *buttons = new QHBoxLayout;
```

Comment:
It creates all the stuff that's supposed to go into the grid, then instantiates the grid and sticks all the elements in there in one block. That's the opposite of interleaving it?

Reply:
The current order of operations is:

1. Create buttons layout
2. Add all the buttons to the buttons layout
3. Create the top-level layout
4. Add widgets to the top-level layout
5. Add buttons layout to the top-level layout

My suggestion would be instead:

1. Create the top-level layout
2. Add widgets to the top-level layout
3. Create buttons layout
4. Add buttons to the buttons layout
5. Add buttons layout to the top-level layout

So that all the buttons-related code is in one block.

Reply:
Understood, I don't wanna do it like that because it would tear apart the grid being constructed, which is already difficult enough to comprehend when the numbers are all together. CurveWidget does it the same way for the same reason (and actually is also doing the extraneous `setLayout` thing so I guess I can patch that up for consistency.)

Reply:
OK. Would it help to use a row counter, or a builder interface that makes the construction look more structured?

I will probably end up making some builder for the other settings layouts since there is a lot of repetition, but the design I am implementing right now mostly uses form layouts and not grid layouts so it wouldn’t probably be too helpful right now for this even if it existed already.

Reply:
Since it's exactly two grid layouts for the sake of dealing with the noodly case of a graph with labels around it, I don't think it's worth a generic solution like a builder or something, the way a graph is presented is probably not going to change anytime soon. Complex grid layouts are otherwise nonexistent, it's either just table-esque grids with an occasional column span of 2 or form layouts, so it's not like we'd get to use that code anywhere else either.